pennylane/devices/qubit/preprocess.py

```diff
@@ -0,0 +1,275 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""This module contains functions for preprocessing `QuantumScript`s to ensure
+that they are supported for execution by a device."""
+
+import pennylane as qml
+
+# from pennylane.tape import QuantumScript
+from pennylane.operation import Observable, Tensor
+from pennylane.measurements import (
+    MidMeasureMP,
+    ExpectationMP,
+    ShadowExpvalMP,
+    SampleMP,
+    ProbabilityMP,
+    CountsMP,
+)
+from pennylane import DeviceError
+
+
+def _stopping_condition(op):
+    """Specify whether or not an Operator object is supported by the device"""
+    return getattr(op, "has_matrix", False) and qml.matrix(op).shape[-1] % 2 == 0
+
+
+def _supports_observable(dev, observable):
+    """Checks if an observable is supported by this device. Raises a ValueError,
+        if not a subclass or string of an Observable was passed.
+
+    Args:
+        dev (.Device): device to check
+        observable (type or str): observable to be checked
+
+    Returns:
+        bool: ``True`` iff supplied observable is supported
+    """
+    if isinstance(observable, type) and issubclass(observable, Observable):
+        observable = observable.__name__
+    if isinstance(observable, str):
+        return observable in dev.observables
+
+    return False
+
+
+def expand_fn(circuit, dev, max_expansion=10):
+    """Method for expanding or decomposing an input circuit. Can be the default or
+    a custom expansion method, see :meth:`.Device.custom_expand` for more details.
+
+    By default, this method expands the tape if:
+
+    - nested tapes are present,
+    - any operations are not supported on the device, or
+    - multiple observables are measured on the same wire.
+
+    Args:
+        circuit (.QuantumTape): the circuit to expand.
+        dev (.Device): the device to execute circuit(s) on.
+        max_expansion (int): The number of times the circuit should be
+            expanded. Expansion occurs when an operation or measurement is not
+            supported, and results in a gate decomposition. If any operations
+            in the decomposition remain unsupported by the device, another
+            expansion occurs.
+
+    Returns:
+        .QuantumTape: The expanded/decomposed circuit, such that the device
+        will natively support all operations.
+    """
+    # pylint: disable=protected-access
+
+    if dev.custom_expand_fn is not None:
+        return dev.custom_expand_fn(circuit, max_expansion=max_expansion)
```

Comment:
So the point of the `preprocess` method is to create a quantum script that the device can natively support without error.

We will allow users to choose compilation steps and user-controlled preprocessing, but we don't want to end up in situations where someone sends something to `execute` that will raise errors, especially errors after a huge amount of computational effort has already been done.

So we will no longer allow dynamically overriding device preprocessing behaviour. Users will either be able to inherit from the device and override the `preprocess` method, or apply their own compilation transforms before we call `preprocess`.